src/data/list/chain.lean

```diff
@@ -203,4 +204,49 @@ begin
     },
 end
 
+/--
+If `a` and `b` are related by the reflexive transitive closure of `r`,
+then there is a chain starting from `a` and ending on `b`./
+
+NB: This is easier to express with `list.chain'` and `head_of_ne_nil` than
+`list.chain` because of the `list.last` required for the end.
+-/
+lemma exists_zigzag {r : α → α → Prop} {a b : α} (h : relation.refl_trans_gen r a b) :
+  ∃ (l : list α), chain' r l ∧ ∃ (hl : l ≠ nil), head_of_ne_nil l hl = a ∧ last l hl = b :=
+begin
+  apply relation.refl_trans_gen.head_induction_on h,
+  { exact ⟨[b], chain.nil, cons_ne_nil _ _, rfl, rfl⟩ },
+  { intros c d e t ih,
+    obtain ⟨l, hl₁, hl₂, rfl, hl₄⟩ := ih,
+    refine ⟨c :: l, _, cons_ne_nil _ _, rfl, _⟩,
+    { cases l,
+      { apply chain'_singleton },
+      { rw chain'_cons,
+        exact ⟨e, hl₁⟩ } },
+    { rwa last_cons _ hl₂ } },
+end
+
+/--
+Given a chain from `a` to `b`, and a predicate true at `b`, if `r x y → p y → p x` then
+the predicate is true at `a`.
+That is, we can propagate the predicate up the chain.
+-/
+lemma prop_up_chain' {r : α → α → Prop} (p : α → Prop) {a b : α}
+  (l : list α) (hl : l ≠ []) (h : chain' r l)
+  (ha : head_of_ne_nil l hl = a) (hb : last l hl = b)
+  (carries : ∀ {x y : α}, r x y → p y → p x) (final : p b) : p a :=
```

Comment:
This I don't know how to name. Let's see if we can summon the naming oracle: @digama0!

Reply:
Looks like `chain'.induction` to me

Reply:
In this and the other statement, you could just use `head l = some a` instead of `head_of_ne_nil`. Better yet, just use `a::l` instead of `l` and then the head is built in, and you can use a function `tail' a l` that gets the tail of `a::l` (does this exist?). That way you can skip the nonempty assumption.

Reply:
I adapted your second suggestion here - it makes everything a lot nicer! Thanks!